modules/rtm/R/invert.auto.R

```diff
@@ -53,25 +51,58 @@
 #' @param return.samples Include full samples list in output. Default = TRUE.
 #' @param save.samples Filename for saving samples after each iteration. If 
 #' 'NULL', do not save samples. Default = NULL.
+#' @param parallel Logical. Whether or not to run multiple chains in parallel on multiple cores (defualt=FALSE).
+#' @param parallel.cores Number of cores to use for parallelization. If NULL (default), allocate one fewer than detected number of cores.
 #' @param ... Other arguments to `check.convergence`
 #' @return List of "results" (summary statistics and Gelman Diagnostic) and 
 #' "samples"(mcmc.list object, or "NA" if return.samples=FALSE)
 
-invert.auto <- function(observed, invert.options, return.samples=TRUE, save.samples=NULL, quiet=FALSE, ...){
+invert.auto <- function(observed, invert.options, return.samples=TRUE, save.samples=NULL, quiet=FALSE, parallel=FALSE, parallel.cores=NULL, ...){
     library(coda)
     n.tries <- invert.options$n.tries
     nchains <- invert.options$nchains
     inits.function <- invert.options$inits.function
     invert.options$do.lsq <- invert.options$do.lsq.first
+    if(invert.options$do.lsq) library(minpack.lm)
     try.again <- TRUE
     i.try <- 1
     while(try.again & i.try <= n.tries){
         print(sprintf("Attempt %d of %d", i.try, n.tries))
-        samps.list <- list()
-        for(chain in 1:nchains){
-            print(sprintf("Chain %d of %d", chain, nchains))
-            invert.options$inits <- inits.function() 
-            samps.list[[chain]] <- invert.custom(observed=observed, invert.options=invert.options, quiet=quiet)
+        if(parallel & !require(parallel)){
+            warning("'parallel' package not installed. Proceeding without parallelization")
+            parallel <- FALSE
+        }
+        if(parallel){
+            library(parallel)
+            invert.function <- function(x){
+                set.seed(x)
+                invert.options$inits <- inits.function()
+                samps <- invert.custom(observed=observed, invert.options=invert.options, quiet=quiet)
+                return(samps)
+            }
+            maxcores <- detectCores()
+            if(is.null(parallel.cores)){
+                cl <- makeCluster(maxcores - 1, "FORK")
+            } else {
+                if(!is.numeric(parallel.cores) | parallel.cores %% 1 != 0){
+                    stop("Invalid argument to 'parallel.cores'. Must be integer or NULL")
+                } else if (parallel.cores > maxcores){
+                    warning(sprintf("Requested %1$d cores but only %2$d cores available. Using only %2$d cores.", parallel.cores, maxcores))
+                    parallel.cores <- maxcores
+                }
+                cl <- makeCluster(parallel.cores, "FORK")
```

Comment:
The package will happily consume all of the cores it can, but if it does so, that renders your computer virtually unusable for anything else (since it has no CPU left to do anything). The default behavior is to leave 1 core for browsing, etc., but if you're passing lots of cores, I think it's safe to assume you want it to perform inversions and nothing else (e.g. running on the cluster). I personally think that's sensible behavior, but if others disagree, it's an easy fix.